test/common/performance/sender/sender.go

```diff
@@ -0,0 +1,266 @@
+		},
+		acceptedEvents: &pb.EventsRecord{
+			Type:   pb.EventsRecord_ACCEPTED,
+			Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+		},
+		failedEvents: &pb.EventsRecord{
+			Type:   pb.EventsRecord_FAILED,
+			Events: make(map[string]*timestamp.Timestamp, estimatedNumberOfTotalMessages),
+		},
+
+		aggregatorClient: aggregatorClient,
+	}
+
+	executor.loadGenerator, err = loadGeneratorFactory(eventsSource(), executor.sentCh, executor.acceptedCh, executor.failedCh)
+	if err != nil {
+		return nil, err
+	}
+
+	return executor, nil
+}
+
+func (s *Sender) Run(ctx context.Context) {
+	// --- Warmup phase
+	log.Printf("--- BEGIN WARMUP ---")
+	if s.warmupSeconds > 0 {
+		if err := s.warmup(ctx, s.warmupSeconds); err != nil {
+			log.Fatalf("Failed to run warmup: %v", err)
+		}
+	} else {
+		log.Printf("Warmup skipped")
+	}
+	log.Printf("---- END WARMUP ----")
+
+	log.Printf("--- BEGIN BENCHMARK ---")
+
+	// Start the events processor
+	log.Println("Starting events processor")
+	go s.processEvents()
+
+	// Clean mess before starting
+	runtime.GC()
+
+	log.Println("Starting benchmark")
+
+	// Run all pace configurations
+	benchmarkBeginning := time.Now()
+	for i, pace := range s.paceSpecs {
+		log.Printf("Starting pace %d° at %v rps for %v seconds", i+1, pace.Rps, pace.Duration)
+		s.loadGenerator.RunPace(i, pace, s.msgSize)
+
+		// Wait for flush
+		time.Sleep(waitForFlush)
+
+		// Trigger GC
+		log.Println("Triggering GC")
+		s.loadGenerator.SendGCEvent()
+		runtime.GC()
+
+		// Wait for receivers GC
+		time.Sleep(waitForReceiverGC)
+	}
+
+	s.loadGenerator.SendEndEvent()
+
+	log.Printf("Benchmark completed in %v", time.Since(benchmarkBeginning))
+
+	s.closeChannels()
+
+	log.Println("---- END BENCHMARK ----")
+
+	log.Println("Sending collected data to the aggregator")
+
+	log.Printf("%-15s: %d", "Sent count", len(s.sentEvents.Events))
+	log.Printf("%-15s: %d", "Accepted count", len(s.acceptedEvents.Events))
+	log.Printf("%-15s: %d", "Failed count", len(s.failedEvents.Events))
+
+	err := s.aggregatorClient.Publish(&pb.EventsRecordList{Items: []*pb.EventsRecord{
+		s.sentEvents,
+		s.acceptedEvents,
+		s.failedEvents,
+	}})
+	if err != nil {
+		log.Fatalf("Failed to send events record: %v\n", err)
+	}
+}
+
+func (s *Sender) warmup(ctx context.Context, warmupSeconds uint) error {
+	log.Println("Starting warmup")
+
+	s.loadGenerator.Warmup(common.PaceSpec{Rps: warmupRps, Duration: time.Duration(warmupSeconds) * time.Second}, s.msgSize)
+
+	// give the channel some time to drain the events it may still have enqueued
+	time.Sleep(5 * time.Second)
+
+	return nil
+}
+
+// processVegetaResult processes the results from the Vegeta attackers.
+func (s *Sender) closeChannels() {
+	log.Printf("All requests sent")
+
+	close(s.sentCh)
+	close(s.acceptedCh)
+	close(s.failedCh)
+
+	// assume all responses are received after a certain time
+	time.Sleep(8 * time.Second)
```

Comment:
This should be a constant

Reply:
This one is not required anymore since now the image waits for the end signal